src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs

```diff
@@ -19,1319 +18,736 @@
 
-            Dictionary<NuGetFramework, RuntimeGraph> resolvedRuntimeGraphs = new();
+            // A LibraryDependency representing the root of the graph which is the project itself
+            LibraryDependency mainProjectDependency = new(
+                new LibraryRange()
+                {
+                    Name = _request.Project.Name,
+                    VersionRange = new VersionRange(_request.Project.Version),
+                    TypeConstraint = LibraryDependencyTarget.Project | LibraryDependencyTarget.ExternalProject
+                });
+
+            // Create the list of framework/runtime pairs to resolve graphs for.  This method returns the pairs in order with the target framework pairs without runtime identifiers come first
+            List<FrameworkRuntimePair> projectFrameworkRuntimePairs = RestoreCommand.CreateFrameworkRuntimePairs(_request.Project, runtimeIds: RequestRuntimeUtility.GetRestoreRuntimes(_request));
 
-            foreach (FrameworkRuntimePair pair in projectFrameworkRuntimePairs.NoAllocEnumerate())
+            // Loop through the target framework and runtime identifier pairs to resolve each graph.
+            // The pairs are sorted with all of the RID-less framework pairs first
+            foreach (FrameworkRuntimePair frameworkRuntimePair in projectFrameworkRuntimePairs.NoAllocEnumerate())
             {
-                if (!string.IsNullOrWhiteSpace(pair.RuntimeIdentifier) && !hasInstallBeenCalledAlready)
+                // Since the FrameworkRuntimePair objects are sorted, the packages found so far need to be installed before resolving any runtime identifier specific graphs because
+                // the runtime.json is in the package which is used to determine what runtime packages to add
+                if (!string.IsNullOrWhiteSpace(frameworkRuntimePair.RuntimeIdentifier) && !hasInstallBeenCalledAlready)
                 {
                     downloadDependencyResolutionResults = await ProjectRestoreCommand.DownloadDependenciesAsync(_request.Project, context, _telemetryActivity, telemetryPrefix: string.Empty, token);
 
-                    _success &= await projectRestoreCommand.InstallPackagesAsync(uniquePackages, allGraphs, downloadDependencyResolutionResults, userPackageFolder, token);
+                    success &= await projectRestoreCommand.InstallPackagesAsync(installedPackages, allGraphs, downloadDependencyResolutionResults, userPackageFolder, token);
 
+                    // This ensures that the packages for the RID-less graph are only installed once
                     hasInstallBeenCalledAlready = true;
                 }
 
-                TargetFrameworkInformation? projectTargetFramework = _request.Project.GetTargetFramework(pair.Framework);
+                // Get the corresponding TargetFrameworkInformation from the restore request
+                TargetFrameworkInformation? projectTargetFramework = _request.Project.GetTargetFramework(frameworkRuntimePair.Framework);
```

Comment:
This should never be null. 
If it is null, the projectFramewokrRuntimePairs generation is wrong. 

Alternatively the framework runtime pairs, can include this type.

Reply:
Yeah that's fair. 

My suggestion is actually just fail if projectTargetFramework is null. 
It means something is really messed up.

Reply:
nvm you did the nullable ignore thing. 

Resolving.